// @basetime/bldr-sfmc-sdk/dist/sfmc/utils/sfmcContextMapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sfmc_context_mapping = void 0;
const sfmc_context_mapping = [
    {
        name: 'Data Extensions',
        rootName: 'Data Extensions',
        context: 'dataExtension',
        contentType: 'dataextension',
    },
    {
        name: 'Shared Data Extensions',
        rootName: 'Shared Data Extensions',
        context: 'sharedDataExtension',
        contentType: 'shared_dataextension',
    },
    {
        name: 'Content Builder',
        rootName: 'Content Builder',
        context: 'contentBuilder',
        contentType: 'asset',
    },
    {
        name: 'Shared Content',
        rootName: 'Shared Content',
        context: 'sharedContent',
        contentType: 'asset',
    },
    {
        name: 'Automation Studio',
        rootName: 'my automations',
        context: 'automationStudio',
        contentType: 'automations',
    },
    {
        name: 'Query',
        rootName: 'Query',
        context: 'automationStudio',
        contentType: 'queryactivity',
        api: 'queries',
    },
    {
        name: 'Scripts',
        rootName: 'Scripts',
        context: 'automationStudio',
        contentType: 'ssjsactivity',
        api: 'scripts',
    },
];
exports.sfmc_context_mapping = sfmc_context_mapping;
